Sources/ProtobufferCoding/Encoding/UnkeyedEncodingContainer.swift

```diff
@@ -0,0 +1,103 @@
+//
+// This source file is part of the Apodini open source project
+//
+// SPDX-FileCopyrightText: 2019-2021 Paul Schmiedmayer and the Apodini project authors (see CONTRIBUTORS.md) <paul.schmiedmayer@tum.de>
+//
+// SPDX-License-Identifier: MIT
+//
+
+import NIO
+import ApodiniUtils
+import Foundation
+
+
+struct ProtobufferUnkeyedEncodingContainer: UnkeyedEncodingContainer {
+    let codingPath: [CodingKey]
+    let dstBufferRef: Box<ByteBuffer>
+    let context: EncoderContext
+    
+    private(set) var count: Int = 0
+    
+    init(codingPath: [CodingKey], dstBufferRef: Box<ByteBuffer>, context: EncoderContext) {
+        self.codingPath = codingPath
+        self.dstBufferRef = dstBufferRef
+        self.context = context
+    }
+    
+    mutating func encodeNil() throws {
+        fatalError("Not implemented")
```

Comment:
(Same applies to the `SingleValueEncodingContainer`, as well as the `UnkeyedDecodingContainer`, as well as the `SingleValueDecodingContainer`)

---

Having these crash instead of throwing errors was a deliberate decision.

The thing about en/decoding protobuffer messages is that you should , in theory, only ever need the `Keyed{Encoding,Decoding}Container`s, and never the SingleValue- or Unkeyed- containers. The reason being that all data in protubuffer messages is encoded using a key-value format. (This includes things like arrays, which, in other formats (e.g. JSON), are the primary use case for the unkeyed coding containers.)

There are some edge cases, however, where, in order to implement the en/decoding of proto messages, it is necessary to fall back back to the SingleValue/Unkeyed containers (e.g.: you're decoding a custom type, have already consumed the field's key, and are now invoking the type's `init(from:)` initializer. The type may then call `.singleValueContainer()` on the decoder, in order to read its value from that.) The important thing about this is that accessing a SingleValue/Unkeyed Container will only ever happen within the context of en/decoding a key-value field.

Meaning that if you end up in one of these crashing functions, something went horribly wrong (since you should never end up there, based on how the proto format is defined), and there really isn't a good way to recover from that (apart from just ignoring what you were trying to en/decode).

We could of course replace it with errors, although tbh the way i see it throwing an error should be used in cases where we were unable to en/decode the data due to some runtime problem, and just straight-up crashing should be used in cases where you should never ever end up in that function in the first place.

One more thing: The proto en/decoders do in fact, before en/decoding something, check whether the type can be represented as a proto message, and throw if that is not the case ([see e.g. here]( https://github.com/Apodini/Apodini/blob/00069cb634b7ee7c4d5db8ad54bdf39c264a89c9/Sources/ProtobufferCoding/Encoding/Encoder.swift#L30)), meaning that we do effectively already throw in these edge cases, and the fatalErrors serve as an indication that something in the proto type validation is implemented incorrectly. Hence why I preferred crashing, since that guarantees that you'll actually be made aware of the issue, otherwise it'd be too easy to just silently ignore the error somewhere.

@PSchmiedmayer what are your thoughts on this?

Reply:
Can this fatal error be triggered by an incoming Protocol Buffer message after the web service did pass compile time and startup time checks?
If it can potentially be reached (e.g. by a malformed message) at runtime then we should throw and error and make sure that this is propagated to the client. Otherwise this would be a very easy attack vector so kill server instances. So this would be hard yes for proper error handling.

If this is logically impossible to reach I can see that a fatal error could be a valid option. Nevertheless we must assume our check logic is implemented correctly this is probably hard to verify? The question would be if this is any harm to properly throw an error here and where would this error send up? Could it potentially be passed to the client?

Reply:
@PSchmiedmayer 

> Can this fatal error be triggered by an incoming Protocol Buffer message after the web service did pass compile time and startup time checks?

Technically speaking this is completely independent from the web service, since the ProtobufferEncoder could be used from anywhere. In reality, client requests sent to a gRPC service are (apart from the tests) the only source of incoming proto messages.

Whether or not these `fatalError`s are reached is completely independent from the actual message received by the web service (or the data being passed to the decoder), since the actual program execution path within the en/decoder (i.e. which functions are called, which if/etc branches are taken, etc) is determined solely by the concrete `Codable` implementation of the Swift type being en/decoded.
Meaning that if we end up reaching one of these fatalErrors, its not because the input data was invalid, but rather because the type being en/decoded has an invalid `Codable` implementation[^1].

If you prefer throwing errors instead, I'll change it to that, though tbh I don't see how a web service (or rather Apodini operating w/ an invalid web service) could reasonably recover from these errors, since ending up in one of these functions would mean that either no incoming requests could ever be decoded, or that all outgoing responses would fail to encode. At which point the web service would already be effectively unusable.


> The question would be if this is any harm to properly throw an error here and where would this error send up? Could it potentially be passed to the client?

I think this is a different question entirely. Whether or not the error reaches the client depends on where exactly it's thrown. From how I understand the code, this would actually be fully outside the control of either ApodiniGRPC or ApodiniProtobuffer: The decoding of a gRPC input message happens in the gRPC IE's `GRPCEndpointParameterDecodingStrategy`, which gets called from somewhere w/in Apodini. So the question of whether these errors could ever reach a client would be the same as whether e.g. JSON decoding errors in the REST IE could ever propagate back to the client.


[^1]: To provide an example of this: If a type's `encode(to:)` function were to be called w/ a `ProtobufEncoder`, and this type then were to try to encode its properties into this encoder's unkeyedEncodingContainer, that'd result in an invalid proto output, because all proto fields need to be key-value pairs. So it's not like the issue would be that it failed to encode this time, but might work at some point in the future, but rather the issue is that this should never be possible at all. And also, the en/decoder will only try to en/decode if the validation succeesed (which throws if it fails), so this should already be taken care of.

Reply:
@lukaskollmer Thanks for the clarifications! 
Based on your insights I would be fine with sticking with fatalErrors and but would suggest to improve the error message with a hing why this is an fatalError and maybe a request to the developer to file a bug in this project if the error would ever be encountered.
Encoding errors would be able to reach the client but as you mentioned that it is impossible for a any specific client request to reach the web service and trigger this function this is probably also not an error message that a client should be able to see and would be useful for a client to see.